spqlios/cplx/cplx_conversions.c

```diff
@@ -0,0 +1,158 @@
+#include <errno.h>
+#include <string.h>
+
+#include "../commons_private.h"
+#include "cplx_fft.h"
+#include "cplx_fft_private.h"
+
+EXPORT void cplx_from_znx32_ref(const CPLX_FROM_ZNX32_PRECOMP* precomp, void* r, const int32_t* x) {
+  const uint32_t m = precomp->m;
+  const int32_t* inre = x;
+  const int32_t* inim = x + m;
+  CPLX* out = r;
+  for (uint32_t i = 0; i < m; ++i) {
+    out[i][0] = (double)inre[i];
+    out[i][1] = (double)inim[i];
+  }
+}
+
+EXPORT void cplx_from_tnx32_ref(const CPLX_FROM_TNX32_PRECOMP* precomp, void* r, const int32_t* x) {
+  static const double _2p32 = 1. / (1L << 32);
+  const uint32_t m = precomp->m;
+  const int32_t* inre = x;
+  const int32_t* inim = x + m;
+  CPLX* out = r;
+  for (uint32_t i = 0; i < m; ++i) {
+    out[i][0] = ((double)inre[i]) * _2p32;
+    out[i][1] = ((double)inim[i]) * _2p32;
+  }
+}
+
+EXPORT void cplx_to_tnx32_ref(const CPLX_TO_TNX32_PRECOMP* precomp, int32_t* r, const void* x) {
+  static const double _2p32 = (1L << 32);
+  const uint32_t m = precomp->m;
+  double factor = _2p32 / precomp->divisor;
+  int32_t* outre = r;
+  int32_t* outim = r + m;
+  const CPLX* in = x;
+  // Note: this formula will only work if abs(in) < 2^32
```

Comment:
We could check that error in the reference implementation, but we will not be able to do this accurately in advanced assembly implementations, so that would result in different behaviour on different architectures. In some sense, the notion of "correct" result is actually not very well defined for an overhead >= 32 anyways.

On the other hand, for most applications, the bounds come from a mathematical guarantee on the use-case: for instance, in TFHE, the overhead is  proved to be below < 2^14.

To maintain some user-awareness, the corresponding `init` function has a `log2overhead` parameter where the user commits to the maximal range of the input, and the API stipulates that any input outside of that committed range is undefined behaviour. -> it allows to select the best implementation choice based on the committed range, and it seems more useful this way.

Reply:
Thanks for the explanation! this makes a lot of sense